File: python/ir/codegen.py
import ffi

def codegen(ast, target: ffi.Target):
	if target.type() == ffi.TargetType.CPU:
		return ffi.code_gen_cpu(ast)
	elif target.type() == ffi.TargetType.GPU:
		return ffi.code_gen_cuda(ast)
	else:
		assert False, "Unrecognized target %s" % target

File: src/codegen/code_gen.cc
#include <codegen/code_gen.h>
#include <codegen/code_gen_cpu.h>
#include <codegen/code_gen_cuda.h>

namespace freetensor {

NativeCode codeGen(const Func &func, const Ref<Target> &target) {
    switch (target->type()) {
    case TargetType::CPU:
        return codeGenCPU(func, target);
#ifdef FT_WITH_CUDA
    case TargetType::GPU:
        return codeGenCUDA(func, target);
#endif // FT_WITH_CUDA
    default:
        ERROR("Unrecognized target " + target->toString());
    }
}

} // namespace freetensor
